unittest/mysys/dynamic_array-t.c

```diff
@@ -0,0 +1,235 @@
+    return EXIT_FAILURE;
+  }
+
+  tmp_thing.id= 1;
+  snprintf(tmp_thing.name, sizeof(tmp_thing.name), "thing 1");
+  err= push_dynamic(&things_array, &tmp_thing);
+  ok(!err, "push_dynamic for 1");
+  if (err) {
+    delete_dynamic(&things_array);
+    return EXIT_FAILURE;
+  }
+
+  ok(2 == things_array.elements, "size expect: %d, actual: %zu",
+     2, things_array.elements);
+
+  popped= (struct thing *)pop_dynamic(&things_array);
+  ok(popped != NULL, "pop_dynamic 1");
+  if (!popped) {
+    delete_dynamic(&things_array);
+    return EXIT_FAILURE;
+  }
+  ok(popped->id == 1, "pop expect 1, popped->id: %u", popped->id);
+  ok(1 == things_array.elements, "size expect: %d, actual: %zu",
+     1, things_array.elements);
+
+  popped= (struct thing *)pop_dynamic(&things_array);
+  ok(popped != NULL, "pop_dynamic 0");
+  if (!popped) {
+    delete_dynamic(&things_array);
+    return EXIT_FAILURE;
+  }
+  ok(popped->id == 0, "pop expect 0, popped->id: %u", popped->id);
+  ok(0 == things_array.elements, "size expect: %d, actual: %zu",
+     0, things_array.elements);
+
+
+  popped= (struct thing *)pop_dynamic(&things_array);
+  ok(!popped, "pop %p from empty array", popped);
+
+  delete_dynamic(&things_array);
+
+  return EXIT_SUCCESS;
+}
+
+struct string_thing {
+  unsigned id;
+  char *str;
+  size_t str_size;
+};
+
+static void free_string_thing(void *p)
+{
+  struct string_thing *thing= (struct string_thing *)p;
+  my_free(thing->str);
+}
+
+int dyn_array_delete_with_callback(void)
+{
+  DYNAMIC_ARRAY things_array;
+  PSI_memory_key psi_key= PSI_NOT_INSTRUMENTED;
+  size_t element_size= sizeof(struct string_thing);
+  size_t init_alloc= 3;
+  size_t alloc_increment= 2;
+  myf my_flags= MYF(MY_WME);
+  my_bool err= FALSE;
+
+  err= my_init_dynamic_array(psi_key, &things_array, element_size, init_alloc,
+                             alloc_increment, my_flags);
+
+  ok(!err, "my_init_dynamic_array");
+  if (err) {
+    return EXIT_FAILURE;
+  }
+
+  for (size_t i= 0; i < 3; ++i) {
+    struct string_thing thing= { i, NULL, 40 };
+    thing.str= my_malloc(psi_key, thing.str_size, my_flags);
+    ok(thing.str != NULL, "%zu thing.str= malloc(%zu))", i, thing.str_size);
+    if (!thing.str) {
+      delete_dynamic_with_callback(&things_array, free_string_thing);
+      return EXIT_FAILURE;
+    }
+    snprintf(thing.str, thing.str_size, "thing %zu", i);
+
+    err= insert_dynamic(&things_array, &thing);
+    ok(!err, "insert_dynamic for %zu", i);
+    if (err) {
+      delete_dynamic_with_callback(&things_array, free_string_thing);
+      return EXIT_FAILURE;
+    }
+  }
+  ok(3 == things_array.elements, "size expect: %d, actual: %zu",
+     3, things_array.elements);
+
+  delete_dynamic_with_callback(&things_array, free_string_thing);
+
+  ok(0 == things_array.elements, "size expect: %d, actual: %zu",
+     0, things_array.elements);
+
+  return EXIT_SUCCESS;
+}
+
+int main(void)
+{
+  int err= 0;
+
+  plan((3 + (2 * (THING_STACK_BUF_SIZE - 1)) + 1)
```

Comment:
I think above the calls to the functions rather than above the functions themselves, since the plan values are variable, but yes, I can think I can make this easier to follow.